internal/mage/util/engine.go

```diff
@@ -0,0 +1,312 @@
+	mergedOpts := map[string]string{}
+	for _, opt := range opts {
+		maps.Copy(mergedOpts, opt.EntrypointArgs)
+	}
+	keys := maps.Keys(mergedOpts)
+	sort.Strings(keys)
+
+	var entrypoint string
+
+	type entrypointTmplParams struct {
+		Bridge            string
+		EngineBin         string
+		EngineConfig      string
+		EntrypointArgs    map[string]string
+		EntrypointArgKeys []string
+	}
+	tmpl := template.Must(template.New("entrypoint").Parse(engineEntrypointTmpl))
+	buf := new(bytes.Buffer)
+	err := tmpl.Execute(buf, entrypointTmplParams{
+		EngineBin:         "/usr/local/bin/" + engineBinName,
+		EngineConfig:      engineTomlPath,
+		EntrypointArgs:    mergedOpts,
+		EntrypointArgKeys: keys,
+	})
+	if err != nil {
+		panic(err)
+	}
+	entrypoint = buf.String()
+
+	return entrypoint, nil
+}
+
+func getConfig(opts ...DevEngineOpts) (string, error) {
+	mergedOpts := map[string]string{}
+	for _, opt := range opts {
+		maps.Copy(mergedOpts, opt.ConfigEntries)
+	}
+	keys := maps.Keys(mergedOpts)
+	sort.Strings(keys)
+
+	var config string
+
+	type configTmplParams struct {
+		ConfigEntries map[string]string
+		ConfigKeys    []string
+	}
+	tmpl := template.Must(template.New("config").Parse(engineConfigTmpl))
+	buf := new(bytes.Buffer)
+	err := tmpl.Execute(buf, configTmplParams{
+		ConfigEntries: mergedOpts,
+		ConfigKeys:    keys,
+	})
+	if err != nil {
+		panic(err)
+	}
+	config = buf.String()
+
+	return config, nil
+}
+
+func CIDevEngineContainerAndEndpoint(ctx context.Context, c *dagger.Client, opts ...DevEngineOpts) (*dagger.Container, string, error) {
+	devEngine := CIDevEngineContainer(c, opts...)
+
+	endpoint, err := devEngine.Endpoint(ctx, dagger.ContainerEndpointOpts{Port: 1234, Scheme: "tcp"})
+	if err != nil {
+		return nil, "", err
+	}
+	return devEngine, endpoint, nil
+}
+
+var DefaultDevEngineOpts = DevEngineOpts{
+	EntrypointArgs: map[string]string{
+		"network-name": "dagger-dev",
+		"network-cidr": "10.88.0.0/16",
+	},
+	ConfigEntries: map[string]string{
+		"grpc":                 `address=["unix:///var/run/buildkit/buildkitd.sock", "tcp://0.0.0.0:1234"]`,
+		`registry."docker.io"`: `mirrors = ["mirror.gcr.io"]`,
+	},
+}
+
+func CIDevEngineContainer(c *dagger.Client, opts ...DevEngineOpts) *dagger.Container {
+	engineOpts := []DevEngineOpts{}
+
+	engineOpts = append(engineOpts, DefaultDevEngineOpts)
+	engineOpts = append(engineOpts, opts...)
+
+	devEngine := devEngineContainer(c, runtime.GOARCH, engineOpts...)
+	devEngine = devEngine.WithExposedPort(1234, dagger.ContainerWithExposedPortOpts{Protocol: dagger.Tcp}).
+		// TODO: in some ways it's nice to have cache here, in others it may actually result in our tests being less reproducible.
+		// Can consider rm -rfing this dir every engine start if we decide we want a clean slate every time.
+		// It's important it's a cache mount though because otherwise overlay won't be available
+		// WithMountedCache("/var/lib/dagger", c.CacheVolume("dagger-dev-engine-state")).
```

Comment:
Just FYI that using a cache mount is strictly required in order for the engine to be able to use overlayfs, which is necessary for the tests to pass (I think they are still passing because this still gets set in the engine test mage step).

Even for the non-test steps we should set this up because use of the fallback native snapshotter instead of overlay will result in everything being very slow and disk space growing n^2 (each "layer" on disk is actually its contents plus a copy of every layer below it).

What was the issue you were seeing around this?

Also, in retrospect the TODO comment I added above this isn't necessary, I was overthinking. Leaving the cache mount around is totally fine and actually exactly what we were doing previously with the dev engine (except there it was a docker volume).

Reply:
For buildkitd.lock, make sure that if there's multiple definitions of the engine they use different cache volume names; there is one instance of a service per hash of the `Container` definition. But ideally we only need one definition so the engine can be shared by everything.

Actually, another even more likely explanation is the service leak issue that's been fixed in main but not released yet. Once it's released we can update the stable engine to use that.

If all that doesn't work and the lock is leaking elsewhere somehow, then we can also just `rm` it when the engine container starts (just as part of the test code, not the actual entrypoint of the engine we publish).